test/core/transport/chaotic_good/client_transport_test.cc

```diff
@@ -260,8 +199,146 @@ class ClientTransportTest : public ::testing::Test {
   MockEndpoint* data_endpoint_ptr_;
   size_t initial_arena_size = 1024;
   MemoryAllocator memory_allocator_;
-  Sequence control_endpoint_sequence;
-  Sequence data_endpoint_sequence;
+  Sequence control_endpoint_write_sequence;
+  Sequence control_endpoint_read_sequence;
+  Sequence data_endpoint_write_sequence;
+  Sequence data_endpoint_read_sequence;
+
+  void AddWriteExpectations(int successful_write_messages,
+                            bool failed_at_last) {
+    if (successful_write_messages > 0) {
+      // Transport started writes.
+      for (int i = 1; i <= successful_write_messages; i++) {
+        EXPECT_CALL(control_endpoint_, Write)
+            .InSequence(control_endpoint_write_sequence)
+            .WillOnce(Return(true));
+        EXPECT_CALL(data_endpoint_, Write)
+            .InSequence(data_endpoint_write_sequence)
+            .WillOnce(Return(true));
+      }
+    }
+    if (failed_at_last) {
+      EXPECT_CALL(control_endpoint_, Write)
+          .InSequence(control_endpoint_write_sequence)
+          .WillOnce(
+              WithArgs<0>([](absl::AnyInvocable<void(absl::Status)> on_write) {
+                on_write(absl::InternalError("control endpoint write failed."));
+                return false;
+              }));
+      EXPECT_CALL(data_endpoint_, Write)
+          .InSequence(data_endpoint_write_sequence)
+          .WillOnce(
+              WithArgs<0>([](absl::AnyInvocable<void(absl::Status)> on_write) {
+                on_write(absl::InternalError("data endpoint write failed."));
+                return false;
+              }));
+    }
+  }
+  void AddReadExpectations(int stream_id, int successful_read_messages,
+                           bool failed_at_last) {
+    if (successful_read_messages > 0) {
```

Comment:
restructure this so that:

if (successful_read_messages <= 0) { return; }
Rest of the body

Reply:
As discussed offline, I moved the error test case to a different file, and simplify the expectations.